crates/matrix-sdk/src/send_queue.rs

```diff
@@ -746,6 +787,162 @@ impl QueueStorage {
+                        Ok(c) => c,
+                        Err(err) => {
+                            warn!("unable to deserialize: {err}");
+                            return Ok(());
+                        }
+                    };
+
+                    // 2. …and a room message, at this point.
+                    let AnyMessageLikeEventContent::RoomMessage(room_message_content) = content
+                    else {
+                        warn!("trying to send an edit event for a non-room message: aborting");
+                        return Ok(());
+                    };
+
+                    // Assume no relation.
+                    let new_content: RoomMessageEventContentWithoutRelation =
+                        room_message_content.into();
+
+                    let edit_event = match room
+                        .make_edit_event(&event_id, EditedContent::RoomMessage(new_content))
+                        .await
+                    {
+                        Ok(e) => e,
+                        Err(err) => {
+                            warn!("couldn't create edited event: {err}");
+                            return Ok(());
+                        }
+                    };
+
+                    // Queue the edit event in the send queue 🧠.
+                    let serializable = SerializableEventContent::from_raw(
+                        Raw::new(&edit_event)
+                            .map_err(RoomSendQueueStorageError::JsonSerialization)?,
+                        edit_event.event_type().to_string(),
+                    );
+
+                    store
+                        .save_send_queue_event(&self.room_id, TransactionId::new(), serializable)
+                        .await
+                        .map_err(RoomSendQueueStorageError::StorageError)?;
+                } else {
+                    // Still local (sending must have failed); update the local echo.
+                    let edited = store
+                        .update_send_queue_event(&self.room_id, &de.transaction_id, new_content)
+                        .await
+                        .map_err(RoomSendQueueStorageError::StorageError)?;
+
+                    if !edited {
+                        warn!("missing local echo upon dependent edit");
+                    }
+                }
+            }
+
+            DependentQueuedEventKind::Redact => {
+                if let Some(event_id) = de.event_id {
+                    // Has been sent; send a redaction.
+                    let room = client
+                        .get_room(&self.room_id)
+                        .ok_or(RoomSendQueueError::RoomDisappeared)?;
+
+                    // Ideally we'd use the send queue to send the redaction, but the protocol has
+                    // changed the shape of a room.redaction after v11, so keep it simple and try
+                    // once here.
+
+                    if let Err(err) = room.redact(&event_id, None, None).await {
```

Comment:
It's not possible to redact messages with a reason if we're using the send queue? That sounds like a Element-caused hole we have here.

Reply:
How is it Element-caused?

And no, not possible to redact; note this is because we're trying to abort a thing that was being sent, and that was not supposed to, so this carries the value of "cancelling the sending", after it happened. Hence, no reasons.

Reply:
Ah duh, I was wrongly imagining that we're not implementing this because Element doesn't need it.